### ISSUE TITLE
Data.x

### ISSUE THREAD

**Issue by humanpose1** [nicolas-chaulet/deeppointcloud-benchmarks#177]
Right now, the way the features are added is very dataset-dependent (in shapenet, we add the normals.
[here](https://github.com/nicolas-chaulet/deeppointcloud-benchmarks/blob/4b70a107ed352b789abce9b044a4b52b078b6a41/src/datasets/segmentation/shapenet.py#L162)
, in s3dis we add the colors and the elevation [here](https://github.com/nicolas-chaulet/deeppointcloud-benchmarks/blob/4b70a107ed352b789abce9b044a4b52b078b6a41/src/datasets/segmentation/s3dis.py#L520).

For KPConv, we add a column of ones directly in the set_input [here](https://github.com/nicolas-chaulet/deeppointcloud-benchmarks/blob/4b70a107ed352b789abce9b044a4b52b078b6a41/src/models/segmentation/kpconv.py#L85) 

It seems a bit unorganized... 
Like transform or pre_transform, It would be great to create classes, that add features we want directly on our dataset.
```python
class MyFeature:
    r"""add a great feature
    Args:
    param: parameter for our feature
    """
    def __init__(self, param):
        self.param = param
    def __call__(self, data):
        # replace add_feature by anything you want
        if(data.x is not None):
            data.x = torch.cat([data.x, add_feature(param)], axis=-1)
        else:
            data.x = add_feature(param)
        return data
```
 So in the yaml, we just need to specify the features we want to add.
The advantage is that we can choose if we want to test S3DIS (or shapenet) with normals or with colors or both or none of them.
Moreover, if we want to add new handcrafted features (curvature, planarity, sphericity...), It would be easier.

What do you think about it ?

**Comment by nicolas-chaulet** [comment]
That's a very good idea. I guess the best way to add it at the moment is to implement transforms. Transforms are then defined in the dataset settings. 
The problem with adding normals or colors with a generic transform is that it is dataset dependent (in shapenet normals are raw[3:6], in s3dis colors are raw[3:6] and elevation is raw[2]). any idea how we could go about that? 
The adding ones thing is certainly easy though and can be done with a trasnform instead of in the set_input.

**Comment by humanpose1** [comment]
Yes, the idea is when we read data in shapenet, we store raw[3:6] in the attribute norm (for s3dis, we store the raw[3:6] in the attribute color). 
And Feature will simply do something like:
```python
class RGBFeature(object):
    """
    add color as feature if it exists
    """

    def __call__(self, data):
        assert hasattr(data, 'color')
        color = data.color
        if data.x is None:
            data.x = color
        else:
            data.x = torch.cat([data.x, color], -1)
        return data
```
or 
```python
class NormalFeature(object):
    """
    add normal as feature. if it doesn't exist, compute normals
    using PCA
    """
    def __call__(self, data):
        if data.norm is None:
            raise NotImplementedError("TODO: Implement normal computation")

        norm = data.norm
        if data.x is None:
            data.x = norm
        else:
            data.x = torch.cat([data.x, norm], -1)
        return data
```

**Comment by nicolas-chaulet** [comment]
👍 
Sounds great indeed! It would clean the datasets quite abit

**Comment by nicolas-chaulet** [comment]
@humanpose1 I think we can close it now, the functionality is here, let's use it! Thanks a lot for the PR!
